### ISSUE TITLE
L2 regularization might be buggy

### ISSUE THREAD

**Issue by Oneplus** [clab/dynet#159]
Hi, seeing that dynet uses global decay to implement sparse l2 regularizer. great work! But there seems still some problem.

To my understanding, the `L2WeightDecay` represents the `s_t` in the sparse sgd update methods:
![image](https://cloud.githubusercontent.com/assets/745499/20225536/c0e06e62-a7f8-11e6-9194-586400be2fd3.png)

If that is right, the `L2WeightDecay.update_weight_decay()` seems forget the `\gamma_t` in the above formula because the decay is only updated with `1-\lambda`. See https://github.com/clab/dynet/blob/master/dynet/weight-decay.h#L24-L29.

What's more, all the trainers in dynet (SimpleSGD, Adam, etc.) use the same decay method. But to my understanding, different optimization algorithms might have different decay factor. Taking adam for example:

![image](https://cloud.githubusercontent.com/assets/745499/20226667/d01df908-a7fd-11e6-891b-a840da377398.png)

Please advise if I am wrong about the regularizer.

**Comment by neubig** [comment]
I haven't checked the math carefully on this but I suspect you are probably right. There are other issues with optimizers (https://github.com/clab/dynet/issues/70) so we may have to re-think the way things are implemented.

**Comment by Oneplus** [comment]
Hi, the formula on sparse sgd is directly copied from the sgd tricks paper.
So there shouldn't be problems in that part. For the Adam part, my
derivation is problematic because the \hat(v) is related with historical
theta so it's not sparse update at all.

I found it difficult for the adaptive sgd algorithms to have any sparse sgd
update form because they record the history. there might be some
approximated way or we just ignore l2 for sparse stuff.

**Comment by vene** [comment]
It's possible to do correct sparse updates for adagrad, the original paper discusses it in context of l1 regularization in [section 5.1](http://www.jmlr.org/papers/volume12/duchi11a/duchi11a.pdf). Two example implementations including both l1 and l2 (for the two different forms of adagrad, namely composite mirror descent vs primal-dual SG) are available in [lazygrad](https://github.com/timvieira/lazygrad) and [lightning](https://github.com/scikit-learn-contrib/lightning/blob/master/lightning/impl/adagrad_fast.pyx#L35).

I haven't yet encountered sparse updates for ADAM, but I would be very interested. From a code perspective, I think it's impossible to do regularization at the optimizer level without having a different implementation for each optimizer. Of course, we can always apply regularization inefficiently directly to the objective.

**Comment by neubig** [comment]
We had an offline discussion about this, and while it is possible to modify the way we do training in order to ensure that L2 functions correctly with these optimizers, we would have to make significant changes to the programming interface. While this bug is not fixed, we made three changes:
- Renamed "L2" to "weight decay" in the command line option to emphasize that this is not L2.
- Added some documentation: http://dynet.readthedocs.io/en/latest/unorthodox.html
- Noted that if you really want correct L2 weight decay, you can take the norm of the parameters and add to the loss function.

So I'll label this as "wontfix" and close, but hopefully this makes things clear.
